Issue #3473: ParameterNameCheck: new scope and excludeScope properties

agcuda opened the pull request:
Issue: #3473 

Add scope and excludeScope properties to ParameterNameCheck

Issue #3381 should make the warnings at the following lines disappear:
- [63:24](https://github.com/checkstyle/checkstyle/pull/3480/files#diff-a0d5f5022546877d69441fab461bfd64R63)
- [77:44](https://github.com/checkstyle/checkstyle/pull/3480/files#diff-a0d5f5022546877d69441fab461bfd64R77)

agcuda commented:
I will add ITs, with a warning, as suggested in PR #3472 since it can't be tested with current google config

romani commented:
1)
> calculateScope

please do single return point.

2) this change is big, I will required regression testing.
While I am doing code review , please review https://github.com/checkstyle/contribution/tree/master/checkstyle-tester we need to run this Check on real code to make sure there is no exception.
we need to run on guava code and verify that that all missed violations are correct. It is partly automated process by https://github.com/checkstyle/contribution/tree/master/patch-diff-report-tool .

Do not hesitate to ask any questions, we are here to help you.

agcuda commented:
Thank you, I will look into it and share the report.
I've just implemented the two suggestions above, keeping the commits separated for your convenience. Please tell me when you'd like them to be squashed.

agcuda commented:
I did a few runs with variations of the rule:
```
 <module name="ParameterName">
      <property name="format" value="^[^d].*$"/>
      <property name="ignoreOverridden" value="false"/>
  </module>
```
Apparently  `scope` and `excludeScope` work as expected, but unfortunately this patch changes the default behaviour. Looking at the [diffs](https://agcuda.github.io/csreps/res-no-d-diff/), parameters of methods of anonymous classes are not checked any more. Lowering the scope to anoninner like this:
```
 <module name="ParameterName">
      <property name="format" value="^[^d].*$"/>
      <property name="scope" value="anoninner"/>
      <property name="ignoreOverridden" value="false"/>
    </module>
```
fixes the problem. Therefore, I think I will change the default to "anoninner" and run more tests.
BTW in the [documentation of Scope](http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/api/Scope.html) I see the value `Scope.NOTHING`: is it something meaningful in this context or I can safely ignore it?

romani commented:
Please squash commits , to let review all from scratch.

agcuda commented:
I have fixed the problem described in by comment above and squashed the commits. Here are the results of the regression runs on guavac:
1. [Original code, from master branch](https://agcuda.github.io/csreps/res-no-d-master/site/)
2. [This PR, no scope, no excludeScope](https://agcuda.github.io/csreps/res-a-no-d/site)
3. [This PR, scope=pubic](https://agcuda.github.io/csreps/res-a-no-d-public/site)
4. [This PR, excludeScope=pubic](https://agcuda.github.io/csreps/res-a-no-d-nonpublic/site)
Run 1 and 2 [gave the same results](https://agcuda.github.io/csreps/res-a-no-d-diff), as expected. The sum of "violations" of run 3 and run 4 give the number of violations of run 2, again as expected.

romani commented:
thanks a lot , result are good, 
but lets make reports more close to real problem validate desirable result right now.

http://checkstyle.sourceforge.net/google_style.html#a5.2.6

https://github.com/checkstyle/checkstyle/blob/25a37e5049b7816b34c552899841a978efc37a63/src/main/resources/google_checks.xml#L105
```
        <module name="ParameterName">
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
            <message key="name.invalidPattern"
             value="Parameter name ''{0}'' must match pattern ''{1}''."/>
</module>
```

we have right now 1044 violations at guava - http://checkstyle.sourceforge.net/reports/google-style/guava/ 

in your test please use config from google config without scopes and with with scopes as you plan to do in final upgrade to google config.

please use https://github.com/checkstyle/contribution/tree/master/patch-diff-report-tool to generate diff report. We need to review all changes in diff report and make sure all changes are expected and your guava report will be more correct.

agcuda commented:
I did a test run, with the [current config](https://agcuda.github.io/csreps/res-a-google/my_check.xml):
```
<module name="ParameterName">
<property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
<message key="name.invalidPattern" value="Parameter name ''{0}'' must match pattern ''{1}''."/>
</module>
```
and with the [new Google config](https://agcuda.github.io/csreps/res-a-google-2/my_check.xml) I'd like to introduce:
```
<module name="ParameterName">
<property name="format" value="^[a-z][a-zA-Z0-9]*$"/>
<property name="excludeScope" value="public"/>
<message key="name.invalidPattern" value="Parameter name ''{0}'' must match pattern ''{1}''."/>
</module>
<module name="ParameterName">
<property name="format" value="^[a-z][a-zA-Z0-9]+$"/>
<property name="scope" value="public"/>
<message key="name.invalidPattern" value="Public parameter name ''{0}'' must match pattern ''{1}''."/>
</module>
```
The [result](https://agcuda.github.io/csreps/res-a-google-diff/index.html) is pretty much what I expected: 
- Most errors disappear because they refer to private methods;
- The remaining errors change from 'Parameter name...' to 'Public parameter name... ', and this means that they match the first rule
- Almost everywhere the violation is caused by a one-char name, however in [some cases](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A960) the problem lies in the fact that the second char is uppercase. I don't know why this is wrong, even with the old specs, however these errors disapper once again because the second rule tolerates them.

I checked a few errors to make sure that they actually make sense and so far it seems they do. I'll check some more; in the meantime, please feel free to tell me if there's some more tests you would like me to do.

romani commented:
355 diff is not tha big.
I did rough division at middle of report.
Please review all violations before guava-tests/test/com/google/common/eventbus/outside/AnnotatedSubscriberFinderTests.java , i will review for all items after this file.

Please be attentive, it is very easy to miss regression.

agcuda commented:
I reviewed the top half, and I found some false negatives:
- [MonitorBasedPriorityBlockingQueue.java:159](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A909)
- [MonitorBasedArrayBlockingQueue.java:210](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A890)
- [ClassSanityTesterTest.java:1109](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A675)
- [ArbitraryInstancesTest.java:429](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A622)
- [DerivedComparable.java:28](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A556)
- [BaseComparable.java:32](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A553)
- [ConcurrentHashMap.java:52](https://agcuda.github.io/csreps/res-a-google-diff/index.html#A327)

They are all public constructors of public classes with a one-char parameter: they should have been reported, but they weren't. I will look into it.

agcuda commented:
I think I have fixed the problem above: a new run with the fix catches the missing violations.
Here is the [diff report](https://agcuda.github.io/csreps/res-ctor-diff/) between the fix and the code in this PR. When you complete the review in the bottom half, if there are no more false positives/negatives, I'll push the fix.

romani commented:
@agcuda , I see a problem with all missed violations of "guava/guava-tests/test/com/google/common/eventbus/outside/AnnotatedSubscriberFinderTests.java"

First example: https://agcuda.github.io/csreps/res-a-google-diff/xref/home/ubuntu/git/checkstyle-fork/contribution/checkstyle-tester/src/main/java/guava/guava-tests/test/com/google/common/eventbus/outside/AnnotatedSubscriberFinderTests.java.html#L71
it is public method (does not matter what class visibility) so it have to be non single character. So it is not expected that such violation is missed.

additional recommendation:
there are number of changes only in messages.
"Public parameter name 's' must match pattern '^[a-z][a-zA-Z0-9]+$'." vs " 	Parameter name 's' must match pattern '^[a-z][a-z0-9][a-zA-Z0-9]*$'."  for public methods (as we do not expect changes there), could be viewed at https://agcuda.github.io/csreps/res-a-google-diff/index.html#A1048 .
Please try to generate one more report to make message the same in both versions of code. http://checkstyle.sourceforge.net/config.html#Custom_messages . It will minimize diff and let us clearly see a regression without distractions.

agcuda commented:
@romani `notASubscriber` is a public method of a private class (`BaseSubscriberFinderTest.Subscriber`), so it is treated like a private method. I believe this is consistent with `JavadocMethodCheck`, which applies a similar logic.

Regarding the error message: I can force the two messages to be the same, but please note that there is a difference between the two regexps: the old one does not allow e.g. `nThreads`, while the new one allows it. I don't know why the old one is so restrictive, so I think it should be fixed by #3381.

romani commented:
> notASubscriber is a public method of a private class (BaseSubscriberFinderTest.Subscriber), so it is treated like a private method. I believe this is consistent with JavadocMethodCheck, which applies a similar logic.

First of all JavadocMethodCheck is unreliable Check, need to be rewritten completely (to use new parser model), too much problems with it.
Secondly, in google style there is no info about visibility to outside of file , look at 
https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names .
Nested classes should be designed in such a way to be easily moved to be top level classes by one click in IDE.
I hate all single character names, they should not exist almost at all.
Lets be as close to Java guide as possible.


>  I don't know why the old one is so restrictive, so I think it should be fixed by #3381.

it is so restrictive to kill the most simple old-school notation on type marker in name, like "iCount" ,"pValue", "aMyArray" (https://en.wikipedia.org/wiki/Hungarian_notation) . C/C++ naming should stay in C/C++ code. Engineers need to stop moving that to java.

agcuda commented:
> First of all JavadocMethodCheck is unreliable Check, need to be rewritten completely (to use new parser 
> model), too much problems with it.
I see: this patch was my modeled after JavadocMethodCheck because it has the same properties. Is there any check that already implements the new parser model and checks the scope of methods, so that I can recast the PR using the state-of-the-art model?

> it is so restrictive to kill the most simple old-school notation on type marker in name, like "iCount" 
>,"pValue", "aMyArray" (https://en.wikipedia.org/wiki/Hungarian_notation) . C/C++ naming should stay in 
> C/C++ code. Engineers need to stop moving that to java.
I am no fan of hungarian notation as well. Thanks for the clarification: since I thought google_checks was meant to verbatim implement the "bare" google coding standards I could not understand the origin of this deviation. If so, I will take care to reproduce the same behaviour on the private method pattern as well.

romani commented:
@agcuda ,

> Is there any check that already implements the new parser model and checks the scope of methods, so that I can recast the PR using the state-of-the-art model?

old javadoc parser model was based on RegExp, new model is ANTLR based - http://checkstyle.sourceforge.net/writingjavadocchecks.html , it is matter only if you try to a Check that validate Javadoc. It is a bit unrelated to Java check. We can not grab from it behavior , because we did not do through testing of it. It exists as is.
It is hard to find similar Check by property ..... just another checking that care about private modifiers:    
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java#L110
And I like this behavior, demands for class should be the same without consideration of its location, as one day it will go out to be top class.  

> I could not understand the origin of this deviation.

Any company could have any written document about code standard, but if there is no tool behind it to enforce it, nobody follow it.
Writing standards is very hard, too much nuances need to be considered. There are a lot of uncertainties in this document. We even tried to report them to Google, but looks like they do not care.
Google have this document but it does no have a tool that enforce it. Each team in Google have slightly different style. Some of them use checkstyle some do not at all, and just use eclipse formatter.
FYI, Our issue against guava: https://github.com/google/guava/issues/1891
Right now style guide is hosted on github, so you can open issue on them to clarify that point.

----------------------

lets update private scope calculation to be more strict and regenerate report and continue PR acceptance.

agcuda commented:
I have changed the logic, and the [results](https://agcuda.github.io/csreps/res-new-acc-diff/) are promising: if the CI checks succeed, I'll start reviewing the diffs from the first file up to AnnotatedSubscriberFinderTests.java.

agcuda commented:
I did a review of all the violations up to UnicodeEscaperTest.java (AnnotatedSubscriberFinderTests.java does not contain any violations any more). They seem correct to me.
I looked at some of the violations that the old algorithm used to suppress, and they were all public methods in private scopes.

romani commented:
false-positive https://agcuda.github.io/csreps/res-new-acc-diff/#A257
violation is missed at "protected char[] escape(int c) {".
it is protected method, so accessible from outside of class, so should be named in non-lazy form.
Exception for private arguments exists only to let developer be lazy on names that does not affect anobody else.

I stopped review as few several issues are the same. Please fix and regenerate report.

agcuda commented:
Clearly we have different interpretations of what is a public method, so before I do another  run, please give me a formal definition so I can generate the corresponding configuration.

romani commented:
> https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names
>> One-character parameter names in public methods should be avoided.

Looks like it is mine problem. Google style is mentioning that only in public one-character is not allowed. Smth blocked my mind, ...... I am switching back to review of test report.

romani commented:
report is good, no problems found. I will do code review one more time before merge.

romani commented on src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/ParameterNameTest.java:
```diff
@@ -63,6 +63,14 @@ public void parameterNameTest() throws Exception {
+            // To be removed after issue #3381 is closed
+            "72:31: " + getCheckMessage(checkConfig.getMessages(), MSG_KEY, "p", format),
+            "77:41: " + getCheckMessage(checkConfig.getMessages(), MSG_KEY, "p", format),
+            // To be removed after issue #3381 is closed
```
please remove such comments and place them to github issue description or comment.

agcuda replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java:
```diff
@@ -105,16 +133,54 @@ public void setIgnoreOverridden(boolean ignoreOverridden) {
                 || ast.getParent().getType() == TokenTypes.LITERAL_CATCH
-                || CheckUtils.isReceiverParameter(ast)) {
+                || CheckUtils.isReceiverParameter(ast)
+                || !matchScope(theScope)) {
```
please move `calculateScope(ast)` inside expression to speedup a little bit performance, scope calculation is not always required.

agcuda replied:
Done

romani commented on src/xdocs/config_naming.xml:
```diff
@@ -1179,6 +1179,18 @@ public boolean equals(Object o) {
+            <td>Visibility scope of methods where parameters are not checked.</td>
+            <td><a href="property_types.html#scope">scope</a></td>
+            <td><code>null</code></td>
+          </tr>
```
please add example of usage of these option.
Examples are the most valuable part of specification. We could take as example config that you used in diff report.

agcuda replied:
Done

rnveach replied:
@agcuda The examples you added went in the description where as we have a sub-section specifically for example configurations. Please move your examples to that sub-section and give them a brief 1 liner to describe what the example accomplishes.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java:
```diff
@@ -72,6 +78,12 @@
     private boolean ignoreOverridden;
 
+    /** The visibility scope where methods are checked. */
+    private Scope scope = Scope.ANONINNER;
```
why you selected this value as default ?
I think it should be NOTHING, or null or absent of Optional.

agcuda replied:
Methods are checked if their access modifier is between `scope` (inclusive) and `excludeScope` (exclusive). The default is to "check all", so `scope` is set to the lowest visibility level and `excludeScope` to `null`, meaning "higher than public". 

The algorithm works because `NodeScope.isIn(Scope.ANONINNER)` returns `true` for all the visibility levels. I tried with `NodeScope.isIn(Scope.NONE)` but it does not. Besides I believe that saying in the documentation that the default scope is `NONE` to say that by default all the scopes are checked is a bit confusing.

Setting the default to `null` would work, of course, because then `matchScope` would check the specific value. Again, as a user I find a bit confusing that:
- `scope = null` means "all the visibility scopes"
- `excludeScope = null` means "no visibility scopes"

but if you prefer it this way, it's perfectly fine for me.

agcuda replied:
@romani Given the considerations above, what is the value you'd like to be the default for `scope`?

romani replied:
@agcuda , fyi https://github.com/checkstyle/checkstyle/issues/3511
sorry for mess in API. Please try to avoid ANONINNER as much as possible to simplify our future refactoring during API cleanup.

agcuda replied:
Kept code as it is.  I can't remove this reference to ANONINNER and keep the code as it is because if e.g. I choose PRIVATE as the default scope, then anonymous inner functions are not checked. To avoid using this constant I could use null or NONE with the drawbacks I described a few comments above

rnveach replied:
I also like the idea of `null` being the default and it means all scopes. Until I looked at the `Scope` class, I didn't understand how `ANONINNER` played in to the check. We should add to description and example that `null` for scope means all.

agcuda replied:
@rnveach Let me know if I get it right: is your suggestion to use `null` as a default both for `scope` and `excludeScope`? And the meaning would be:
- `scope=null`: check all scopes
- `excludeScope=null`: exclude no scopes
Is it correct?

rnveach replied:
@agcuda yes that is my suggestion. Otherwise I would think you would have to explain `ANONINNER` to the users. Explaining `null` seems easier to understand.

rnveach replied:
@agcuda After talking with romani, we decided to leave it as `ANONINNER` so we can get this PR finished sooner. I will add this as a note to the referenced issue so it can be fixed there when we remove the enum.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java:
```diff
@@ -105,16 +133,54 @@ public void setIgnoreOverridden(boolean ignoreOverridden) {
 
     @Override
     protected boolean mustCheckName(DetailAST ast) {
+        final Scope theScope = calculateScope(ast);
```
we should not calculate scope if it is not defined by user.
Check need to work as before update on existing configurations.

agcuda commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java:
```diff
@@ -106,15 +134,56 @@ public void setIgnoreOverridden(boolean ignoreOverridden) {
+    private static Scope calculateScope(final DetailAST ast) {
+        final DetailAST params = ast.getParent();
+        final DetailAST meth = params.getParent();
+        Scope scope = Scope.PRIVATE;
```
@romani This is the scope of e.g lambda function parameters. Strictly speaking, the default should have been  'Scope.ANONINNER', however in light of your comments, I chose to use 'Scope.PRIVATE'. Please tell me if that's ok for you.

agcuda commented:
Is there anything I can do to fix the CircleCI error? I don't understand what's amiss, and it does not look like a random error.

rnveach commented:
> cannot initialize module BeforeExecutionExclusionFileFilter

@agcuda Did you rebase on the latest master? This is a new change we just added.

agcuda commented:
@rnveach No, the merge base of this branch is still 7f8309b (Sept 27).

romani commented:
@agcuda , please rebase over latest code (this new module was added at Oct 22, 2016)

romani commented:
@agcuda ,  I am not sure what is status of your changes. Are you going to change smth base on my comments or ... 
please pin me when you are ready for code-review or you need some help from me.

agcuda commented:
@romani I have implemented all the changes you suggested, so the code is ready for a new review. I only have a doubt about the correct scope for lambda function parameters, see my comment to ParameterNameCheck above.

romani commented:
@agcuda , I still see ANONINNER usage in code , can you get rid of it ? 
please reply to each of my comment as "done" or "....."  (all other points are not covered or I miss smth ....)

agcuda commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java:
```diff
@@ -106,15 +134,56 @@ public void setIgnoreOverridden(boolean ignoreOverridden) {
     @Override
     protected boolean mustCheckName(DetailAST ast) {
         boolean checkName = true;
+        final boolean isDefault = scope == Scope.ANONINNER && excludeScope == null;
```
This reference to ANONINNER is necessary to implement your suggestion about not doing anything if the user has not overridden the defaults (the comment disappeared for some reason).

agcuda commented:
@romani Please see comments above

rnveach commented on src/xdocs/config_naming.xml:
```diff
@@ -1145,7 +1145,25 @@ class MyClass {
+          For example, the following configuration checks that the parameters always start with two
+          lowercase characters and, in addition, that public method parameters cannot be one character
+          long:
+          <pre>
```
Examples must be in `<source>` tag, not `<pre>` to be validated by our unit tests.

rnveach commented on src/xdocs/config_naming.xml:
```diff
@@ -1145,7 +1145,25 @@ class MyClass {
+&lt;property name=&quot;format&quot; value=&quot;^[a-z][a-z0-9][a-zA-Z0-9]*$&quot;/&gt;
+    &lt;property name=&quot;scope&quot; value=&quot;public&quot;/&gt;
+    &lt;message key=&quot;name.invalidPattern&quot;
+        value=&quot;Parameter name ''{0}'' must match pattern ''{1}''./&gt;
```
Example of validation not being run against this example. Line is missing end quote to value attribute.

rnveach commented:
@romani My changes were addressed.

romani commented:
reports were rechecked previously, code is ok.
fix is merged.

romani commented:
fix is merged.
